flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSinkITCase.java

```diff
@@ -426,6 +429,26 @@ public void testNullEnforcer() throws ExecutionException, InterruptedException {
         }
     }
 
+    @Test
+    public void testSelfUnionWatermarkPropagation() throws Exception {
+        DataStream<Integer> dataStream = env.fromElements(1, 2, 3);
+
+        WatermarkOperator operator = new WatermarkOperator(env.getParallelism());
+        dataStream
+                .union(dataStream)
+                .transform("Watermark Operator", BasicTypeInfo.INT_TYPE_INFO, operator)
+                .addSink(new DiscardingSink<>());
```

Comment:
This might have been a misunderstanding: What I meant with a test on the table side is an actual table sink, to test the story end to end. Using the original example from the JIRA issue with `fromValues` and our mock table sink.